dependency: fix merging tags in 'merge_repeats'

UniqMartin opened the pull request:
While it may suffice to merge string and non-reserved tags by forming a union of all tags of dependencies of the same name, this approach fails to work for the reserved tags. These are now merged such that the most restrictive tag (meaning sometimes an empty tag) is preserved.

The previous behavior caused essential dependencies to be omitted and builds to fail in response. E.g., multiple `:fortran` dependencies with tags `[]`, `[:recommended]`, and `[:optional]` would have been expanded and merged to `"gcc"` with tags `[:recommended, :optional]`, causing it to be no longer seen as a required dependency.

To reproduce the issue fixed by this PR, start from a clean Homebrew installation and then execute:

```
$ brew tap homebrew/science
$ brew install -v petsc
==> Installing petsc from homebrew/science
==> Installing dependencies for homebrew/science/petsc: sphinx-doc, cmake, superlu43, metis, parmetis, superlu_dist, veclibfort, scalapack, mumps, hypre, sundials, szip, hdf5, pkg-config, hwloc, tbb, suite-sparse, netcdf, fftw
[…]
```

Notice how the list of dependencies to be installed neither contains `gcc` nor `open-mpi`, though [`petsc`](https://github.com/Homebrew/homebrew-science/blob/16c758b2298186330183520e4882dbc173528e3b/petsc.rb#L22-L23) has a mandatory dependency on both due to its `:fortran` and `:mpi` requirement. With this PR applied, the same invocation of `brew install` instead correctly yields:

```
==> Installing petsc from homebrew/science
==> Installing dependencies for homebrew/science/petsc: xz, gmp, mpfr, libmpc, isl, gcc, pkg-config, makedepend, openssl, libevent, open-mpi, sphinx-doc, cmake, superlu43, metis, parmetis, superlu_dist, veclibfort, scalapack, mumps, hypre, sundials, szip, hdf5, hwloc, tbb, suite-sparse, netcdf, fftw
[…]
```

**Notes:** I'm not particularly attached to any newly introduced names, I just couldn't come up with something better than “necessity” and “temporality”. I'd appreciate feedback for better naming. Please also scrutinize the merging logic and its representation in the tests.

(In some sense, this is a follow-up and extension of the fix in #46916.)

mikemcquaid commented on Library/Homebrew/test/test_dependency.rb:
```diff
@@ -48,7 +48,7 @@ def test_merge_repeats
 
     foo_named_dep = merged.find {|d| d.name == "foo"}
-    assert_equal [:build, "bar"], foo_named_dep.tags
+    assert_equal ["bar"], foo_named_dep.tags
```
This definitely seems wrong to me but I'll read more to understand why.

mikemcquaid replied:
Nope, I agree with the overall changes but this seems to be a mistake/regression.

UniqMartin replied:
I'm pretty sure the previous behavior is wrong. If we have a dependency tree that contains two instances of the same dependency, e.g.:

```
a
|-- xz => :build
\-- b
    \-- xz
```

Then `xz` is ultimately required at both build and runtime and thus the two instances of this dependency should be merged to the equivalent of `depends_on "xz"`.

mikemcquaid replied:
Of course! I see now, thanks for explaining.

mikemcquaid commented on Library/Homebrew/dependency.rb:
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
```
I think `none?` might make more sense here.

UniqMartin replied:
It's kinda related to the `FIXME` comment below. Ideally, I would be just using `deps.any?(&:required?)` instead of the block here, but because of the (debatable) definition of `required?` I'm doing it like this (and trying to keep it as close as possible to the definition of `required?`).

mikemcquaid commented on Library/Homebrew/dependency.rb:
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
+      elsif deps.all?(&:run?)
+        [:run]
+      else
+        [] # Implies build & runtime dependency.
```
Mind using `and` instead of `&` just to avoid confusion with the actual `&` operator?

UniqMartin replied:
If this helps avoid confusion, sure.

mikemcquaid commented on Library/Homebrew/dependency.rb:
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
+        # FIXME: Should be `required?`, but `required?` implies `!build?`.
```
Not sure about the `FIXME` here given it doesn't actually apply to this code; maybe just make this a comment and add the `fixme` elsewhere.

UniqMartin replied:
Yes, maybe the `FIXME` should go into the definition of `required?`.

Ideally, our treatment of dependencies would be a bit more uniform than is currently the case (if my assessment of the current situation is correct), resulting in a different definition of `required?` or both `optional?` and `recommended?` in `Dependable`. But that would be a too far-reaching change for this PR (and it requires substantially more testing). This PR on the other hand tries to be a fix that can be merged (hopefully) relatively quickly.

mikemcquaid replied:
> Yes, maybe the FIXME should go into the definition of required?.

Yeh, I think so.

mikemcquaid commented:
Nice work here :+1:

mikemcquaid commented:
Other than the cosmetic changes here I'm :+1: to :ship: (although I suspect @xu-cheng may also want to take a look).

xu-cheng commented on Library/Homebrew/dependency.rb:
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
+        [:recommended]
+      else # deps.all?(&:optional?)
+        [:optional]
+      end
```
I wonder whether it would help to make the code more readable by implementing this function in this way:
```ruby
def merge_necessity(deps)
  recommended = false
  deps.each do |dep|
    if dep.recommended?
      recommended = true
    elsif !dep.optional?
     return []
    end
  end
  recommended ? [:recommended] : [:optional]
end
```

As a bonus, it will only enumerate the list for once.

UniqMartin replied:
I personally think that my implementation is more readable, because it is expressed on a higher level. I find it easier to reason about all dependencies at once instead of understanding the various state changes that happen in the “loop” you suggest. And I think correctness is (in this case) more important than performance.

Here's my implementation translated into English which hopefully illustrates my point:

- If **any** of the dependencies is a **required** dependency:
  - The merged dependency shall be **required**.
- Otherwise, if **any** of the dependencies is a **recommended** dependency:
  - The merged dependency shall be **recommended**.
- Otherwise, **all** dependencies must be **optional** and:
  - The merged dependency shall be **optional**, too.

mikemcquaid replied:
Although @xu-cheng's involves a single loop I personally find @UniqMartin's more readable. I'm not convinced that the performance overhead is going to me enough to justify the drop in readability here.

xu-cheng commented on Library/Homebrew/dependable.rb:
```diff
@@ -23,7 +23,15 @@ def required?
 
+  def build_and_run?
+    !build? && !run?
+  end
```
This method doesn't seem to be used anywhere except in test case. Should we keep it?

UniqMartin replied:
I agree that it is somewhat redundant. I removed it.

xu-cheng commented:
Few small advice. LGTM :+1:

 :heart: to see we are addressing more edge cases .Thanks @UniqMartin

UniqMartin commented:
@mikemcquaid @xu-cheng I hope I addressed all of your comments either in the updated code I just pushed or by giving reasons why I'd prefer to stay with my implementation.

The issue solved here seems to affect more formulae than I expected – especially in homebrew/science, see <https://github.com/Homebrew/homebrew/pull/46461#issuecomment-165128667> –, so I'm tempted to merge this very soon.
